### PR TITLE
Test, enotice fixes, handling for permissions key for Member_Tasks

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Alternate potential approach to the enotice portion of https://github.com/civicrm/civicrm-core/pull/20940

This approach adds support for permissions using a syntax like the [summary actions hook](https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_summaryActions/) when defining a task - ie

```
$task['blah'] => [
  'permissions' => ['edit memberships'],
];

$task['blah2'] => [
  'permissions' => [['edit memberships', 'view memberships]],
];

```
If we get to the point where it is consistent we can document it as an alternative for extensions (although they may still choose to stick with 'unset') when working with the [search tasks hook](https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_searchTasks/)

The last place where we provide task (I hope) is [searchKitTasks](https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_searchKitTasks/) which kinda leverages search tasks

Before
----------------------------------------
No tests on the function, enotices when tasks are unset, confusing code with lots of setting & unsetting and no 'source of knowledge' - ie consistent metadata.

After
----------------------------------------
A more metadata based approach, potential for more cleanup / integration with `parent:permissionedTaskTitles`

image

Technical Details
----------------------------------------
It took me quite a while to figure out what the EDIT_PERMISSION actually meant and I didn't have any good ideas for how to handle it other than extending test cover and documentation and marking the metadata parameter as 'possibly temporary'

Comments
----------------------------------------
@mattwire potential alternate approach to the enotices in the PR you put up.
@colemanw note the permissions key would potentially be one hooks could set in search kit too - although as I mention they ALSO have the option to unset based on more complex conditions

### CODE REVIEW

File: CRM/Core/Task.php [hunk 1]
```diff
@@ -165,6 +165,51 @@ public static function getTask($value) {
...
+   *
+   * @param array $tasks
+   * @param bool $hasEditContactPermission
+   *   Does the user have permission to edit the given contact. Required where
+   *   permission to edit the user is required in conjunction with permission
+   *   to do the task.
+   *
+   * @return array
+   */
+  protected static function getFilteredTiles(array $tasks, bool $hasEditContactPermission): array {
```
Comment (mattwire) [inline CRM/Core/Task.php:204]:
```suggestion
  protected static function getFilteredTitles(array $tasks, bool $hasEditContactPermission): array {
```

File: CRM/Member/Task.php [hunk 2]
```diff
@@ -135,24 +151,8 @@ public static function taskTitles() {
...
-      ];
-      //CRM-4418,
-      if (CRM_Core_Permission::check('delete in CiviMember')) {
-        $tasks[self::TASK_DELETE] = self::$_tasks[self::TASK_DELETE]['title'];
-      }
-    }
-
-    $tasks = parent::corePermissionedTaskTitles($tasks, $permission, $params);
-    return $tasks;
+    $tasks = self::getFilteredTiles(self::$_tasks, $permission === CRM_Core_Permission::EDIT);
```
Comment (mattwire) [inline CRM/Member/Task.php:154]:
```suggestion
    $tasks = self::getFilteredTitles(self::$_tasks, $permission === CRM_Core_Permission::EDIT);
```

File: CRM/Core/Task.php [hunk 3]
```diff
@@ -165,6 +165,51 @@ public static function getTask($value) {
...
+   *
+   * @param array $tasks
+   * @param bool $hasEditContactPermission
+   *   Does the user have permission to edit the given contact. Required where
+   *   permission to edit the user is required in conjunction with permission
+   *   to do the task.
+   *
+   * @return array
+   */
+  protected static function getFilteredTasks(array $tasks, bool $hasEditContactPermission): array {
```
Comment (mattwire) [inline CRM/Core/Task.php:179]: Maybe `getTasksFilteredByPermission()` to be more explicit/clear what this does?

**Comment by mattwire** [comment]
@eileenmcnaughton This seems a nice approach. I've opened #20945 for the actual "crash" fix and happy for #20940 to be closed in favour of this approach which seems much more sustainable.

File: CRM/Core/Task.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Core/Task.php:204]: Yeah that makes sense - on re-reading I realise the calling function says 'titles' not 'tiles' too

File: CRM/Core/Task.php (hunk 3 above)
Comment (eileenmcnaughton) [reply CRM/Core/Task.php]: Yep - updated to that (& the other function to match)

**Pull request merged**